base/special/gamma.jl

```diff
@@ -630,12 +607,98 @@ function eta(z::Union{Float64,Complex{Float64}})
         return -zeta(z) * expm1(0.6931471805599453094172321214581765*δz)
     end
 end
-eta(x::Integer) = eta(Float64(x))
-eta(x::Real)    = oftype(float(x),eta(Float64(x)))
 
-"""
-    eta(x)
 
-Dirichlet eta function ``\\eta(s) = \\sum^\\infty_{n=1}(-1)^{n-1}/n^{s}``.
-"""
-eta(z::Complex) = oftype(float(z),eta(Complex128(z)))
+# Converting types that we can convert, and not ones we can not
+# If we really cared about single precision, we could make a faster
+# Float32 version by truncating the Stirling series at a smaller cutoff.
+# and if we really cared about half precision, we could make a faster
+# Float16 version, by using a precomputed table look-up.
+
+# Float16, and Float32 and their Complex equivalents can be converted to Float64
+# and results converted back. Similar for BitIntegers (eg Int32), but no converting back
+# Otherwise, we need to make things use their own `float` converting methods
+# and in those cases, we do not convert back either as we assume
+# they also implement there own versions of the functions
+
+
+ofpromotedtype(as::Tuple, c) = convert(promote_type(typeof.(as)...), c)
+ComplexOrRealUnion(TS...) = Union{(ComplexOrReal{T} for T in TS)...}
+
+const types_le_Float64 = (Float16, Float32, Float64, Base.BitInteger.types...)
+
+for T in types_le_Float64
+    @eval f64(x::Complex{$T}) = Complex128(x)
+    @eval f64(x::$T) = Float64(x)
+end
+
+
+for f in (:digamma, :trigamma, :zeta, :eta, :invdigamma)
+    @eval begin
+        $f(z::ComplexOrRealUnion(Base.BitInteger.types...)) = $f(f64(z))
+        $f(z::ComplexOrRealUnion(Float16,Float32)) = oftype(z, $f(f64(z)))
+
+        function $f(z::Number)
+            x = float(z)
+            typeof(x) === typeof(z) && throw(MethodError($f, (z,)))
+            # There is nothing to fallback to, since this didn't change the argument types
+            $f(x)
+        end
+    end
+end
+
+function polygamma(m::Integer, z::ComplexOrRealUnion(Float16,Float32))
+    oftype(z, polygamma(m, f64(z)))
+end
+
+for T1 in types_le_Float64, T2 in types_le_Float64
```

Comment:
This is indeed a particularly convoluted piece of metaprogramming.
I rewrote it about 3 times in different ways.
but I think I'll try again.
Get rid of some of the `if T1<:Integer && T2<:Integer`
by using a separate loop for ints.

I have been trying to avoid things like
`zeta(x::BigFloat,y::BigFloat) = error...`, because that will mean that if someone does write a `zeta(::BigFloat, ::BigFloat)`
it will have issues with overwriting a function name. And will trigger #265 problems.

In general, rather than exclude the types in `Base` that a function does not work for,
I am trying to include only the types in Base which a function does work for.
Which should make extending with custom number types from packages sane.